File: ports/php/src/Security/Covert/VariationSelectorPayload.php
<?php

declare(strict_types=1);

namespace UnicodePhp\Security\Covert;

use UnicodePhp\Data;
use UnicodePhp\Security\ClassificationKind;
use UnicodePhp\Security\Crypto\AiWatermarkDetectability;

interface VsSubThreat
{
    public function tag(): string;
}

final class VsDirectPayload implements VsSubThreat
{
    public function __construct(public readonly string $decoded)
    {
    }

    public function tag(): string
    {
        return 'DirectPayload';
    }
}

final class VsIllegalTarget implements VsSubThreat
{
    public function __construct(
        public readonly int $targetCp,
        public readonly int $vsCp,
    ) {
    }

    public function tag(): string
    {
        return 'IllegalTarget';
    }
}

final class VsRepeatedBase implements VsSubThreat
{
    public function __construct(
        public readonly int $baseCp,
        public readonly int $vsCount,
    ) {
    }

    public function tag(): string
    {
        return 'RepeatedBase';
    }
}

/// A registered selector precedes the suspicious run: payload hiding behind a
/// legitimate glyph (Lean embeddedAfterReg).
final class VsEmbeddedAfterRegistered implements VsSubThreat
{
    public function __construct(
        public readonly int $registeredEnd,
        public readonly int $payloadStart,
    ) {
    }

    public function tag(): string
    {
        return 'EmbeddedAfterRegistered';
    }
}

final class VsVerdict
{
    /**
     * @param list<int> $vsPositions every selector position, the census
     * @param list<int> $suspiciousPositions the selectors no registered pair or
     *        presentation rule sanctions: what the classification localises
     * @param list<int> $recoveredBytes
     */
    public function __construct(
        public readonly ClassificationKind $kind,
        public readonly ?VsSubThreat $sub,
        public readonly array $vsPositions,
        public readonly array $suspiciousPositions,
        public readonly array $recoveredBytes,
    ) {
    }
}

/// Detection of GlassWorm-class invisible payloads encoded in Unicode variation
/// selectors. Exempts (base, VS) pairs registered in StandardizedVariants.txt or
/// emoji-variation-sequences.txt per UCD 17.0 / UTS #51.
final class VariationSelectorPayload
{
    /** @var array<string,bool>|null */
    private static ?array $legalPairs = null;

    private static function parseHex(string $s): ?int
    {
        $s = trim($s);
        if ($s === '' || preg_match('/^[0-9A-Fa-f]+$/', $s) !== 1) {
            return null;
        }
        return (int) hexdec($s);
    }

    /** @return array<string,bool> */
    private static function legalPairs(): array
    {
        if (self::$legalPairs !== null) {
            return self::$legalPairs;
        }
        $out = [];
        foreach (['StandardizedVariants.txt', 'emoji-variation-sequences.txt'] as $file) {
            foreach (Data::lines($file) as $rawLine) {
                $hash = strpos($rawLine, '#');
                $body = $hash === false ? $rawLine : substr($rawLine, 0, $hash);
                $stripped = trim($body);
                if ($stripped === '') {
                    continue;
                }
                $semi = strpos($stripped, ';');
                $pairPart = $semi === false ? $stripped : substr($stripped, 0, $semi);
                $tokens = preg_split('/\s+/', trim($pairPart), -1, PREG_SPLIT_NO_EMPTY) ?: [];
                if (count($tokens) < 2) {
                    continue;
                }
                $base = self::parseHex($tokens[0]);
                $vs = self::parseHex($tokens[1]);
                if ($base === null || $vs === null) {
                    continue;
                }
                $out[$base . ',' . $vs] = true;
            }
        }
        self::$legalPairs = $out;
        return $out;
    }

    /// True iff `(base, vs)` is a registered variation sequence.
    public static function isRegisteredVariationPair(int $base, int $vs): bool
    {
        return isset(self::legalPairs()[$base . ',' . $vs]);
    }

    public static function isVariationSelector(int $cp): bool
    {
        return ($cp >= 0xFE00 && $cp <= 0xFE0F)
            || ($cp >= 0xE0100 && $cp <= 0xE01EF)
            || ($cp >= 0x180B && $cp <= 0x180D);
    }

    /// Decode a single VS codepoint to its nibble value in [0, 255], or `null`
    /// for the Mongolian FVS codepoints (180B..180D).
    public static function vsToNibble(int $cp): ?int
    {
        if ($cp >= 0xFE00 && $cp <= 0xFE0F) {
            return $cp - 0xFE00;
        }
        if ($cp >= 0xE0100 && $cp <= 0xE01EF) {
            return $cp - 0xE0100 + 16;
        }
        return null;
    }

    /**
     * @param list<int> $input
     * @param list<int> $positions
     * @return list<int>
     */
    private static function decodeVsRun(array $input, array $positions): array
    {
        $out = [];
        $high = null;
        foreach ($positions as $p) {
            $n = self::vsToNibble($input[$p]);
            if ($n === null) {
                continue;
            }
            if ($high === null) {
                $high = $n;
            } else {
                $out[] = (($high << 4) | $n) & 0xFF;
                $high = null;
            }
        }
        return $out;
    }

    /**
     * @param list<int> $input
     * @param list<int> $positions
     */
    private static function allSameVs(array $input, array $positions): bool
    {
        if ($positions === []) {
            return true;
        }
        $cp0 = $input[$positions[0]];
        foreach ($positions as $p) {
            if ($input[$p] !== $cp0) {
                return false;
            }
        }
        return true;
    }

    /** @param list<int> $bytes */
    private static function lossyAscii(array $bytes): string
    {
        $s = '';
        foreach ($bytes as $b) {
            if (($b >= 0x20 && $b <= 0x7E) || $b === 0x09 || $b === 0x0A || $b === 0x0D) {
                $s .= chr($b);
            } else {
                $s .= '?';
            }
        }
        return $s;
    }

    /**
     * @param list<int> $input
     */
    public static function detect(array $input): VsVerdict
    {
        $vsPositions = [];
        foreach ($input as $i => $cp) {
            if (self::isVariationSelector($cp)) {
                $vsPositions[] = $i;
            }
        }

        // Each selector is judged against its predecessor (Lean classifyVS):
        // registered uses are sanctioned wherever they stand and however many;
        // the hazard is the suspicious run alone.
        $registered = [];
        $suspicious = [];
        foreach ($vsPositions as $p) {
            if (self::isRegisteredVariationUse($input, $p)) {
                $registered[] = $p;
            } else {
                $suspicious[] = $p;
            }
        }
        if ($suspicious === []) {
            return new VsVerdict(ClassificationKind::Clear, null, $vsPositions, [], []);
        }

        $recoveredBytes = self::decodeVsRun($input, $suspicious);
        $payloadStart = $suspicious[0];
        $registeredBefore = null;
        foreach ($registered as $p) {
            if ($p < $payloadStart) {
                $registeredBefore = $p;
            }
        }
        // Priority (Lean pickSubThreat): a registered selector before the
        // suspicious run, then a long single-selector run, then a decodable
        // payload, then the bare illegal target.
        if ($registeredBefore !== null) {
            $sub = new VsEmbeddedAfterRegistered($registeredBefore, $payloadStart);
        } elseif (count($suspicious) >= 4 && self::allSameVs($input, $suspicious)) {
            $base = $payloadStart === 0 ? 0 : $input[$payloadStart - 1];
            $sub = new VsRepeatedBase($base, count($suspicious));
        } elseif ($recoveredBytes !== []) {
            $sub = new VsDirectPayload(self::lossyAscii($recoveredBytes));
        } else {
            $target = $payloadStart === 0 ? 0 : $input[$payloadStart - 1];
            $sub = new VsIllegalTarget($target, $input[$payloadStart]);
        }

        return new VsVerdict(ClassificationKind::Hazard, $sub, $vsPositions, $suspicious, $recoveredBytes);
    }

    /**
     * The Lean classifyVS registered reading of the selector at $p: a
     * standardized or emoji variation sequence, or VS15 / VS16 on any base
     * carrying the Emoji property. A selector with no predecessor is never
     * registered.
     *
     * @param list<int> $input
     */
    public static function isRegisteredVariationUse(array $input, int $p): bool
    {
        if ($p === 0) {
            return false;
        }
        $base = $input[$p - 1];
        $vs = $input[$p];
        return self::isRegisteredVariationPair($base, $vs)
            || (($vs === 0xFE0F || $vs === 0xFE0E) && AiWatermarkDetectability::isEmoji($base));
    }
}

File: ports/python/src/unicode_python/security/covert/variation_selector_payload.py
"""Detection of GlassWorm-class invisible payloads encoded in
Unicode variation selectors.

Threat model.  Tier A1.  Adversary crafts an input consisting of
one visible base codepoint followed by a sequence of variation-
selector codepoints (U+FE00..U+FE0F union U+E0100..U+E01EF) that
the receiving renderer treats as a no-op glyph variant but that
a downstream string-processing layer (e.g. an LLM tokenizer or a
clipboard pipeline) preserves byte-for-byte.  Decoding pairs of
VS codepoints back into bytes recovers an arbitrary payload.

This port treats every variation-selector occurrence after a base
codepoint as suspicious.  The Lean reference additionally exempts
(base, VS) pairs that appear in StandardizedVariants.txt and
emoji-presentation pairs — those exemptions require UCD tables.
The coarser port is more conservative: legitimate math / Mongolian
/ Egyptian / CJK Compat / emoji-presentation VS uses will be
flagged.
"""

from dataclasses import dataclass, field
from pathlib import Path
from typing import Union

from ..calculus import ClassificationKind
from ..crypto.ai_watermark_detectability import is_emoji

# ──────────────────────────────────────────────────────────────────────
# Authoritative legal (base, VS) pair set — UCD StandardizedVariants
# + UTS #51 emoji-variation-sequences.  Loaded lazily on first use.
# ──────────────────────────────────────────────────────────────────────

_DATA_DIR = Path(__file__).resolve().parent.parent.parent / "data"


def _parse_legal_variation_pairs() -> set[tuple[int, int]]:
    out: set[tuple[int, int]] = set()
    for fname in (
        "StandardizedVariants.txt",
        "emoji-variation-sequences.txt",
    ):
        path = _DATA_DIR / fname
        with path.open("r", encoding="utf-8") as f:
            for raw_line in f:
                # Strip comment then take up to first ';'.
                hash_idx = raw_line.find("#")
                body = raw_line if hash_idx < 0 else raw_line[:hash_idx]
                semi_idx = body.find(";")
                pair_part = body if semi_idx < 0 else body[:semi_idx]
                tokens = pair_part.split()
                if len(tokens) < 2:
                    continue
                try:
                    base = int(tokens[0], 16)
                    vs = int(tokens[1], 16)
                except ValueError:
                    continue
                out.add((base, vs))
    return out


_LEGAL_PAIRS: set[tuple[int, int]] | None = None


def is_registered_variation_pair(base: int, vs: int) -> bool:
    """True iff ``(base, vs)`` is a registered variation sequence
    per UCD StandardizedVariants.txt or UTS #51
    emoji-variation-sequences.txt.  Used to exempt legitimate
    math / emoji-presentation variants from IllegalTarget
    false-positives."""
    global _LEGAL_PAIRS
    if _LEGAL_PAIRS is None:
        _LEGAL_PAIRS = _parse_legal_variation_pairs()
    return (base, vs) in _LEGAL_PAIRS


def is_variation_selector(cp: int) -> bool:
    if 0xFE00 <= cp <= 0xFE0F:
        return True
    if 0xE0100 <= cp <= 0xE01EF:
        return True
    if 0x180B <= cp <= 0x180D:
        return True
    return False


def vs_to_nibble(cp: int) -> int | None:
    """Decode a single VS codepoint to its nibble value in [0, 255].
    Uses GlassWorm's bit layout: VS1..VS16 → nibbles 0..15,
    VS17..VS256 → nibbles 16..255.  Mongolian FVS codepoints
    (180B..180D) return ``None``."""
    if 0xFE00 <= cp <= 0xFE0F:
        return cp - 0xFE00
    if 0xE0100 <= cp <= 0xE01EF:
        return cp - 0xE0100 + 16
    return None


@dataclass(frozen=True, slots=True)
class DirectPayload:
    decoded: str


@dataclass(frozen=True, slots=True)
class IllegalTarget:
    target_cp: int
    vs_cp: int


@dataclass(frozen=True, slots=True)
class RepeatedBase:
    base_cp: int
    vs_count: int


@dataclass(frozen=True, slots=True)
class EmbeddedAfterRegistered:
    """A registered selector precedes the suspicious run: payload hiding behind
    a legitimate glyph (Lean ``embeddedAfterReg``)."""

    registered_end: int
    payload_start: int


SubThreat = Union[DirectPayload, IllegalTarget, RepeatedBase, EmbeddedAfterRegistered]


def sub_threat_tag(sub: SubThreat) -> str:
    if isinstance(sub, DirectPayload):
        return "DirectPayload"
    if isinstance(sub, IllegalTarget):
        return "IllegalTarget"
    if isinstance(sub, EmbeddedAfterRegistered):
        return "EmbeddedAfterRegistered"
    return "RepeatedBase"


@dataclass(slots=True)
class Verdict:
    kind: ClassificationKind
    sub: SubThreat | None = None
    # Every variation-selector position: the census.
    vs_positions: list[int] = field(default_factory=list)
    # The selectors no registered pair or presentation rule sanctions: what the
    # classification localises (Lean ``suspiciousPositions``).
    suspicious_positions: list[int] = field(default_factory=list)
    recovered_bytes: bytes = b""


def is_registered_use(input_cps: list[int], p: int) -> bool:
    """The Lean ``classifyVS`` registered reading of the selector at ``p``: a
    standardized or emoji variation sequence, or VS15 / VS16 on any base with
    the Emoji property. A selector with no predecessor is never registered."""
    if p == 0:
        return False
    base = input_cps[p - 1]
    vs = input_cps[p]
    return is_registered_variation_pair(base, vs) or (
        vs in (0xFE0F, 0xFE0E) and is_emoji(base)
    )


def _decode_vs_run(input_cps: list[int], positions: list[int]) -> bytes:
    out = bytearray()
    high: int | None = None
    for p in positions:
        n = vs_to_nibble(input_cps[p])
        if n is None:
            continue
        if high is None:
            high = n
        else:
            out.append(((high << 4) | n) & 0xFF)
            high = None
    return bytes(out)


def _all_same_vs(input_cps: list[int], positions: list[int]) -> bool:
    if not positions:
        return True
    cp0 = input_cps[positions[0]]
    return all(input_cps[p] == cp0 for p in positions)


def _lossy_ascii(payload: bytes) -> str:
    out = []
    for b in payload:
        if (0x20 <= b <= 0x7E) or b in (0x09, 0x0A, 0x0D):
            out.append(chr(b))
        else:
            out.append("?")
    return "".join(out)


def detect(input_cps: list[int]) -> Verdict:
    v = Verdict(kind=ClassificationKind.CLEAR)
    v.vs_positions = [i for i, cp in enumerate(input_cps) if is_variation_selector(cp)]

    # Each selector is judged against its predecessor (Lean ``classifyVS``):
    # registered uses are sanctioned wherever they stand and however many; the
    # hazard is the suspicious run alone.
    registered = [p for p in v.vs_positions if is_registered_use(input_cps, p)]
    suspicious = [p for p in v.vs_positions if not is_registered_use(input_cps, p)]
    if not suspicious:
        return v

    v.recovered_bytes = _decode_vs_run(input_cps, suspicious)
    v.kind = ClassificationKind.HAZARD

    payload_start = suspicious[0]
    # Priority (Lean ``pickSubThreat``): a registered selector before the
    # suspicious run, then a long single-selector run, then a decodable
    # payload, then the bare illegal target.
    registered_before = [p for p in registered if p < payload_start]
    if registered_before:
        v.sub = EmbeddedAfterRegistered(
            registered_end=registered_before[-1], payload_start=payload_start
        )
    elif len(suspicious) >= 4 and _all_same_vs(input_cps, suspicious):
        base = 0 if payload_start == 0 else input_cps[payload_start - 1]
        v.sub = RepeatedBase(base_cp=base, vs_count=len(suspicious))
    elif v.recovered_bytes:
        v.sub = DirectPayload(decoded=_lossy_ascii(v.recovered_bytes))
    else:
        target = 0 if payload_start == 0 else input_cps[payload_start - 1]
        v.sub = IllegalTarget(target_cp=target, vs_cp=input_cps[payload_start])
    v.suspicious_positions = suspicious
    return v


__all__ = [
    "DirectPayload",
    "EmbeddedAfterRegistered",
    "IllegalTarget",
    "RepeatedBase",
    "SubThreat",
    "Verdict",
    "detect",
    "is_variation_selector",
    "sub_threat_tag",
    "vs_to_nibble",
]
